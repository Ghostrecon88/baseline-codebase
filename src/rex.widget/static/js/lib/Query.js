/**
 * Bindings to HTSQL query HTTP API.
 *
 * @copyright 2015, Prometheus Research LLC
 */

export {Query as default} from './data/Query';
