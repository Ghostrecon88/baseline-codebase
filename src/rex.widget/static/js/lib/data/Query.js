/**
 * @copyright 2015, Prometheus Research, LLC
 */

import {fetch} from '../fetch';
import shallowEquals from '../shallowEquals';

export class Query {

  static fetch = fetch;

  constructor(path, params) {
    this.path = path;
    this._params = params;
  }

  params(params) {
    return new this.constructor(this.path, {...this._params, ...params});
  }

  produce(params) {
    let query = {
      ...this._params,
      ...params
    };
    return this.constructor.fetch(this.path, query);
  }

  produceCollection(params) {
    return this.produce(params);
  }

  produceEntity(params) {
    return this.produce(params);
  }

  equals(other) {
    return (
      other &&
      this.constructor === other.constructor &&
      this.path === other.path &&
      shallowEquals(this._params, other._params)
    );
  }
}

export function isQuery(obj) {
  return obj instanceof Query;
}

export default function query(path, params = {}) {
  return new Query(path, params);
}
