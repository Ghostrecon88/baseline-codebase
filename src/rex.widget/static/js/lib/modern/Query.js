/**
 * Bindings to HTSQL query HTTP API.
 *
 * @copyright 2015, Prometheus Research LLC
 */
'use strict';

var request   = require('../request');

class Query {

  constructor(path) {
    this.path = path;
  }

  produce(params) {
    return request('GET', this.path)
      .query(params)
      .set('Accept', 'application/json')
      .promise()
      .then(this._handleResponse);
  }

  produceCollection(params) {
    return this.produce()
  }

  produceEntity(params) {
    return this.produce()
  }
}

module.exports = Query;
