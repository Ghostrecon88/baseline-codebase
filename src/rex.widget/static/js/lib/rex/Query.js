/**
 * @jsx React.DOM
 */
'use strict';

var React = require('react');

var Query = React.createClass({

  render() {
    var {data: {data}} = this.props;
    return <div>{JSON.stringify(data)}</div>;
  }
});

module.exports = Query;
