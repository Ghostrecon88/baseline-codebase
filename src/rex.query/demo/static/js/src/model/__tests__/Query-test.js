import * as q from '../Query';

import {
  navigate, filter, limit,
  select, aggregate, pipeline, def
} from '../Query';
import {
  voidType, numberType, textType,
  entityType, recordType,
  seqType, optType
} from '../Type';

function stripDomain(query) {
  return q.map(query, query => {
    return {
      ...query,
      context: query.context ? {type: query.context.type} : null
    };
  });
}

describe('inferType()', function() {

  let domain = {
    aggregate: {
      count: {
        makeType: _ => numberType
      }
    },
    entity: {
      sample: {
        attribute: {
          title: {
            type: textType,
          }
        }
      },
      individual: {
        attribute: {
          name: {
            type: textType,
          },
          sample: {
            type: seqType(entityType('sample')),
          }
        }
      }
    }
  };

  let individual = navigate('individual');
  let sample = navigate('sample');
  let name = navigate('name');
  let title = navigate('title');
  let count = aggregate('count');

  it('infers type of query atoms', function() {
    expect(q.inferType(domain, individual).context.type)
      .toEqual(
        seqType(entityType('individual'))
      );
    expect(q.inferType(domain, navigate('unknown')).context.type)
      .toEqual(null);
    expect(q.inferType(domain, filter(individual)).context.type)
      .toEqual(voidType);
    expect(q.inferType(domain, limit(10)).context.type)
      .toEqual(voidType);
    expect(q.inferType(domain, select({a: individual})).context.type)
      .toEqual(recordType({
        a: seqType(entityType('individual'))
      }));
    expect(q.inferType(domain, select({a: navigate('unknown')})).context.type)
      .toEqual(recordType({
        a: null
      }));
    expect(q.inferType(domain, aggregate('count')).context.type)
      .toEqual(null);
  });

  it('fails on navigate to unknown attribute', function() {
    expect(q.inferType(domain, pipeline(
      individual,
      navigate('b'),
    )).context.type).toEqual(
      null
    );
  });

  it('fails on navigate to unknown record field', function() {
    expect(q.inferType(domain, pipeline(
      select({a: individual}),
      navigate('b'),
    )).context.type).toEqual(
      null
    );
  });

  it('individual.name', function() {
    let query = pipeline(
      individual,
      name
    );
    expect(stripDomain(q.inferType(domain, query))).toMatchSnapshot();
  });

  it('individual.name:count()', function() {
    let query = pipeline(
      individual,
      name,
      count
    );
    expect(stripDomain(q.inferType(domain, query))).toMatchSnapshot();
  });


  it('individual.sample', function() {
    let query = pipeline(
      individual,
      sample,
    );
    expect(stripDomain(q.inferType(domain, query))).toMatchSnapshot();
  });

  it('individual.sample.title', function() {
    let query = pipeline(
      individual,
      sample,
      title,
    );
    expect(stripDomain(q.inferType(domain, query))).toMatchSnapshot();
  });

  it('individual:select(a := name, b := name)', function() {
    let query = pipeline(
      individual,
      select({
        a: name,
        b: name
      })
    );
    expect(stripDomain(q.inferType(domain, query))).toMatchSnapshot();
  });

  it('individual:select(a := name, s := sample)', function() {
    let query = pipeline(
      individual,
      select({
        a: name,
        s: sample,
      })
    );
    expect(stripDomain(q.inferType(domain, query))).toMatchSnapshot();
  });

  it('individual:select(a := name, s := sample:count())', function() {
    let query = pipeline(
      individual,
      select({
        a: name,
        s: pipeline(sample, count)
      })
    );
    expect(stripDomain(q.inferType(domain, query))).toMatchSnapshot();
  });

  it('individual:select(a := name, s := sample.title)', function() {
    let query = pipeline(
      individual,
      select({
        a: name,
        s: pipeline(sample, title)
      })
    );
    expect(stripDomain(q.inferType(domain, query))).toMatchSnapshot();
  });

  it('individual:select(a := name, b := name).a', function() {
    let query = pipeline(
      individual,
      select({a: name, b: name}),
      navigate('a'),
    );
    expect(stripDomain(q.inferType(domain, query))).toMatchSnapshot();
  });

  it('individual:unknown() : fails on unknown aggregate', function() {
    let query = pipeline(
      individual,
      aggregate('unknown')
    );
    expect(stripDomain(q.inferType(domain, query))).toMatchSnapshot();
  });

  it('individual:define(nickname := name)', function() {
    let query = pipeline(
      individual,
      def('nickname', navigate('name')),
    );
    expect(stripDomain(q.inferType(domain, query))).toMatchSnapshot();
  });

  it('individual:define(nickname := name).nickname', function() {
    let query = pipeline(
      individual,
      def('nickname', navigate('name')),
      navigate('nickname'),
    );
    expect(stripDomain(q.inferType(domain, query))).toMatchSnapshot();
  });

});
