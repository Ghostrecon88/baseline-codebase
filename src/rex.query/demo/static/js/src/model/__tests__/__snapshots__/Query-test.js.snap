exports[`inferType() individual.name 1`] = `
Object {
  "context": Object {
    "type": Object {
      "name": "seq",
      "type": Object {
        "name": "text",
      },
    },
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "name": "text",
          },
        },
      },
      "name": "navigate",
      "path": "name",
    },
  ],
}
`;

exports[`inferType() individual.name:count() 1`] = `
Object {
  "context": Object {
    "type": Object {
      "name": "number",
    },
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "name": "text",
          },
        },
      },
      "name": "navigate",
      "path": "name",
    },
    Object {
      "aggregate": "count",
      "context": Object {
        "type": Object {
          "name": "number",
        },
      },
      "name": "aggregate",
    },
  ],
}
`;

exports[`inferType() individual.sample 1`] = `
Object {
  "context": Object {
    "type": Object {
      "name": "seq",
      "type": Object {
        "entity": "sample",
        "name": "entity",
      },
    },
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "sample",
            "name": "entity",
          },
        },
      },
      "name": "navigate",
      "path": "sample",
    },
  ],
}
`;

exports[`inferType() individual.sample.title 1`] = `
Object {
  "context": Object {
    "type": Object {
      "name": "seq",
      "type": Object {
        "name": "text",
      },
    },
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "sample",
            "name": "entity",
          },
        },
      },
      "name": "navigate",
      "path": "sample",
    },
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "name": "text",
          },
        },
      },
      "name": "navigate",
      "path": "title",
    },
  ],
}
`;

exports[`inferType() individual:define(newname := name):select(a := newname, b := name) 1`] = `
Object {
  "context": Object {
    "type": Object {
      "name": "seq",
      "type": Object {
        "fields": Object {
          "a": Object {
            "name": "text",
          },
          "b": Object {
            "name": "text",
          },
        },
        "name": "record",
      },
    },
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "binding": Object {
        "name": "newname",
        "query": Object {
          "context": Object {
            "type": Object {
              "name": "seq",
              "type": Object {
                "name": "text",
              },
            },
          },
          "name": "pipeline",
          "pipeline": Array [
            Object {
              "context": Object {
                "type": Object {
                  "name": "seq",
                  "type": Object {
                    "name": "text",
                  },
                },
              },
              "name": "navigate",
              "path": "name",
            },
          ],
        },
      },
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "define",
    },
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "fields": Object {
              "a": Object {
                "name": "text",
              },
              "b": Object {
                "name": "text",
              },
            },
            "name": "record",
          },
        },
      },
      "name": "select",
      "select": Object {
        "a": Object {
          "context": Object {
            "type": Object {
              "name": "text",
            },
          },
          "name": "pipeline",
          "pipeline": Array [
            Object {
              "context": Object {
                "type": Object {
                  "name": "text",
                },
              },
              "name": "navigate",
              "path": "newname",
            },
          ],
        },
        "b": Object {
          "context": Object {
            "type": Object {
              "name": "text",
            },
          },
          "name": "pipeline",
          "pipeline": Array [
            Object {
              "context": Object {
                "type": Object {
                  "name": "text",
                },
              },
              "name": "navigate",
              "path": "name",
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`inferType() individual:define(nickname := name) 1`] = `
Object {
  "context": Object {
    "type": Object {
      "name": "seq",
      "type": Object {
        "entity": "individual",
        "name": "entity",
      },
    },
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "binding": Object {
        "name": "nickname",
        "query": Object {
          "context": Object {
            "type": Object {
              "name": "seq",
              "type": Object {
                "name": "text",
              },
            },
          },
          "name": "pipeline",
          "pipeline": Array [
            Object {
              "context": Object {
                "type": Object {
                  "name": "seq",
                  "type": Object {
                    "name": "text",
                  },
                },
              },
              "name": "navigate",
              "path": "name",
            },
          ],
        },
      },
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "define",
    },
  ],
}
`;

exports[`inferType() individual:define(nickname := name).nickname 1`] = `
Object {
  "context": Object {
    "type": Object {
      "name": "seq",
      "type": Object {
        "name": "text",
      },
    },
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "binding": Object {
        "name": "nickname",
        "query": Object {
          "context": Object {
            "type": Object {
              "name": "seq",
              "type": Object {
                "name": "text",
              },
            },
          },
          "name": "pipeline",
          "pipeline": Array [
            Object {
              "context": Object {
                "type": Object {
                  "name": "seq",
                  "type": Object {
                    "name": "text",
                  },
                },
              },
              "name": "navigate",
              "path": "name",
            },
          ],
        },
      },
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "define",
    },
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "name": "text",
          },
        },
      },
      "name": "navigate",
      "path": "nickname",
    },
  ],
}
`;

exports[`inferType() individual:define(nickname := name):select(nickname) 1`] = `
Object {
  "context": Object {
    "type": Object {
      "name": "seq",
      "type": Object {
        "fields": Object {
          "nickname": Object {
            "name": "text",
          },
        },
        "name": "record",
      },
    },
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "binding": Object {
        "name": "nickname",
        "query": Object {
          "context": Object {
            "type": Object {
              "name": "seq",
              "type": Object {
                "name": "text",
              },
            },
          },
          "name": "pipeline",
          "pipeline": Array [
            Object {
              "context": Object {
                "type": Object {
                  "name": "seq",
                  "type": Object {
                    "name": "text",
                  },
                },
              },
              "name": "navigate",
              "path": "name",
            },
          ],
        },
      },
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "define",
    },
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "fields": Object {
              "nickname": Object {
                "name": "text",
              },
            },
            "name": "record",
          },
        },
      },
      "name": "select",
      "select": Object {
        "nickname": Object {
          "context": Object {
            "type": Object {
              "name": "text",
            },
          },
          "name": "pipeline",
          "pipeline": Array [
            Object {
              "context": Object {
                "type": Object {
                  "name": "text",
                },
              },
              "name": "navigate",
              "path": "nickname",
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`inferType() individual:define(nickname := name):select(nickname:select(code)) 1`] = `
Object {
  "context": Object {
    "type": Object {
      "name": "seq",
      "type": Object {
        "fields": Object {
          "nickname": Object {
            "fields": Object {
              "code": null,
            },
            "name": "record",
          },
        },
        "name": "record",
      },
    },
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "binding": Object {
        "name": "nickname",
        "query": Object {
          "context": Object {
            "type": Object {
              "name": "seq",
              "type": Object {
                "name": "text",
              },
            },
          },
          "name": "pipeline",
          "pipeline": Array [
            Object {
              "context": Object {
                "type": Object {
                  "name": "seq",
                  "type": Object {
                    "name": "text",
                  },
                },
              },
              "name": "navigate",
              "path": "name",
            },
          ],
        },
      },
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "define",
    },
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "fields": Object {
              "nickname": Object {
                "fields": Object {
                  "code": null,
                },
                "name": "record",
              },
            },
            "name": "record",
          },
        },
      },
      "name": "select",
      "select": Object {
        "nickname": Object {
          "context": Object {
            "type": Object {
              "fields": Object {
                "code": null,
              },
              "name": "record",
            },
          },
          "name": "pipeline",
          "pipeline": Array [
            Object {
              "context": Object {
                "type": Object {
                  "name": "text",
                },
              },
              "name": "navigate",
              "path": "nickname",
            },
            Object {
              "context": Object {
                "type": Object {
                  "fields": Object {
                    "code": null,
                  },
                  "name": "record",
                },
              },
              "name": "select",
              "select": Object {
                "code": Object {
                  "context": Object {
                    "type": null,
                  },
                  "name": "pipeline",
                  "pipeline": Array [
                    Object {
                      "context": Object {
                        "type": null,
                      },
                      "name": "navigate",
                      "path": "code",
                    },
                  ],
                },
              },
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`inferType() individual:define(s := sample:count()):select(s:select(title)) 1`] = `
Object {
  "context": Object {
    "type": Object {
      "name": "seq",
      "type": Object {
        "fields": Object {
          "s": Object {
            "fields": Object {
              "title": null,
            },
            "name": "record",
          },
        },
        "name": "record",
      },
    },
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "binding": Object {
        "name": "s",
        "query": Object {
          "context": Object {
            "type": Object {
              "name": "number",
            },
          },
          "name": "pipeline",
          "pipeline": Array [
            Object {
              "context": Object {
                "type": Object {
                  "name": "seq",
                  "type": Object {
                    "entity": "sample",
                    "name": "entity",
                  },
                },
              },
              "name": "navigate",
              "path": "sample",
            },
            Object {
              "aggregate": "count",
              "context": Object {
                "type": Object {
                  "name": "number",
                },
              },
              "name": "aggregate",
            },
          ],
        },
      },
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "define",
    },
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "fields": Object {
              "s": Object {
                "fields": Object {
                  "title": null,
                },
                "name": "record",
              },
            },
            "name": "record",
          },
        },
      },
      "name": "select",
      "select": Object {
        "s": Object {
          "context": Object {
            "type": Object {
              "fields": Object {
                "title": null,
              },
              "name": "record",
            },
          },
          "name": "pipeline",
          "pipeline": Array [
            Object {
              "context": Object {
                "type": Object {
                  "name": "number",
                },
              },
              "name": "navigate",
              "path": "s",
            },
            Object {
              "context": Object {
                "type": Object {
                  "fields": Object {
                    "title": null,
                  },
                  "name": "record",
                },
              },
              "name": "select",
              "select": Object {
                "title": Object {
                  "context": Object {
                    "type": null,
                  },
                  "name": "pipeline",
                  "pipeline": Array [
                    Object {
                      "context": Object {
                        "type": null,
                      },
                      "name": "navigate",
                      "path": "title",
                    },
                  ],
                },
              },
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`inferType() individual:select(a := age) 1`] = `
Object {
  "context": Object {
    "type": Object {
      "name": "seq",
      "type": Object {
        "fields": Object {
          "a": Object {
            "name": "opt",
            "type": Object {
              "name": "text",
            },
          },
        },
        "name": "record",
      },
    },
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "fields": Object {
              "a": Object {
                "name": "opt",
                "type": Object {
                  "name": "text",
                },
              },
            },
            "name": "record",
          },
        },
      },
      "name": "select",
      "select": Object {
        "a": Object {
          "context": Object {
            "type": Object {
              "name": "opt",
              "type": Object {
                "name": "text",
              },
            },
          },
          "name": "navigate",
          "path": "age",
        },
      },
    },
  ],
}
`;

exports[`inferType() individual:select(a := name, b := name) 1`] = `
Object {
  "context": Object {
    "type": Object {
      "name": "seq",
      "type": Object {
        "fields": Object {
          "a": Object {
            "name": "text",
          },
          "b": Object {
            "name": "text",
          },
        },
        "name": "record",
      },
    },
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "fields": Object {
              "a": Object {
                "name": "text",
              },
              "b": Object {
                "name": "text",
              },
            },
            "name": "record",
          },
        },
      },
      "name": "select",
      "select": Object {
        "a": Object {
          "context": Object {
            "type": Object {
              "name": "text",
            },
          },
          "name": "navigate",
          "path": "name",
        },
        "b": Object {
          "context": Object {
            "type": Object {
              "name": "text",
            },
          },
          "name": "navigate",
          "path": "name",
        },
      },
    },
  ],
}
`;

exports[`inferType() individual:select(a := name, b := name).a 1`] = `
Object {
  "context": Object {
    "type": Object {
      "name": "seq",
      "type": Object {
        "name": "text",
      },
    },
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "fields": Object {
              "a": Object {
                "name": "text",
              },
              "b": Object {
                "name": "text",
              },
            },
            "name": "record",
          },
        },
      },
      "name": "select",
      "select": Object {
        "a": Object {
          "context": Object {
            "type": Object {
              "name": "text",
            },
          },
          "name": "navigate",
          "path": "name",
        },
        "b": Object {
          "context": Object {
            "type": Object {
              "name": "text",
            },
          },
          "name": "navigate",
          "path": "name",
        },
      },
    },
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "name": "text",
          },
        },
      },
      "name": "navigate",
      "path": "a",
    },
  ],
}
`;

exports[`inferType() individual:select(a := name, s := sample) 1`] = `
Object {
  "context": Object {
    "type": Object {
      "name": "seq",
      "type": Object {
        "fields": Object {
          "a": Object {
            "name": "text",
          },
          "s": Object {
            "name": "seq",
            "type": Object {
              "entity": "sample",
              "name": "entity",
            },
          },
        },
        "name": "record",
      },
    },
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "fields": Object {
              "a": Object {
                "name": "text",
              },
              "s": Object {
                "name": "seq",
                "type": Object {
                  "entity": "sample",
                  "name": "entity",
                },
              },
            },
            "name": "record",
          },
        },
      },
      "name": "select",
      "select": Object {
        "a": Object {
          "context": Object {
            "type": Object {
              "name": "text",
            },
          },
          "name": "navigate",
          "path": "name",
        },
        "s": Object {
          "context": Object {
            "type": Object {
              "name": "seq",
              "type": Object {
                "entity": "sample",
                "name": "entity",
              },
            },
          },
          "name": "navigate",
          "path": "sample",
        },
      },
    },
  ],
}
`;

exports[`inferType() individual:select(a := name, s := sample.title) 1`] = `
Object {
  "context": Object {
    "type": Object {
      "name": "seq",
      "type": Object {
        "fields": Object {
          "a": Object {
            "name": "text",
          },
          "s": Object {
            "name": "seq",
            "type": Object {
              "name": "text",
            },
          },
        },
        "name": "record",
      },
    },
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "fields": Object {
              "a": Object {
                "name": "text",
              },
              "s": Object {
                "name": "seq",
                "type": Object {
                  "name": "text",
                },
              },
            },
            "name": "record",
          },
        },
      },
      "name": "select",
      "select": Object {
        "a": Object {
          "context": Object {
            "type": Object {
              "name": "text",
            },
          },
          "name": "navigate",
          "path": "name",
        },
        "s": Object {
          "context": Object {
            "type": Object {
              "name": "seq",
              "type": Object {
                "name": "text",
              },
            },
          },
          "name": "pipeline",
          "pipeline": Array [
            Object {
              "context": Object {
                "type": Object {
                  "name": "seq",
                  "type": Object {
                    "entity": "sample",
                    "name": "entity",
                  },
                },
              },
              "name": "navigate",
              "path": "sample",
            },
            Object {
              "context": Object {
                "type": Object {
                  "name": "seq",
                  "type": Object {
                    "name": "text",
                  },
                },
              },
              "name": "navigate",
              "path": "title",
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`inferType() individual:select(a := name, s := sample:count()) 1`] = `
Object {
  "context": Object {
    "type": Object {
      "name": "seq",
      "type": Object {
        "fields": Object {
          "a": Object {
            "name": "text",
          },
          "s": Object {
            "name": "number",
          },
        },
        "name": "record",
      },
    },
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "fields": Object {
              "a": Object {
                "name": "text",
              },
              "s": Object {
                "name": "number",
              },
            },
            "name": "record",
          },
        },
      },
      "name": "select",
      "select": Object {
        "a": Object {
          "context": Object {
            "type": Object {
              "name": "text",
            },
          },
          "name": "navigate",
          "path": "name",
        },
        "s": Object {
          "context": Object {
            "type": Object {
              "name": "number",
            },
          },
          "name": "pipeline",
          "pipeline": Array [
            Object {
              "context": Object {
                "type": Object {
                  "name": "seq",
                  "type": Object {
                    "entity": "sample",
                    "name": "entity",
                  },
                },
              },
              "name": "navigate",
              "path": "sample",
            },
            Object {
              "aggregate": "count",
              "context": Object {
                "type": Object {
                  "name": "number",
                },
              },
              "name": "aggregate",
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`inferType() individual:unknown() : fails on unknown aggregate 1`] = `
Object {
  "context": Object {
    "type": null,
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "type": Object {
          "name": "seq",
          "type": Object {
            "entity": "individual",
            "name": "entity",
          },
        },
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "aggregate": "unknown",
      "context": Object {
        "type": null,
      },
      "name": "aggregate",
    },
  ],
}
`;
